File: _archive/src_python_mvp/services/gemini_service.py
import json
import time
import logging
from typing import Dict, Any, Optional
from google import genai
from google.genai import types
from google.genai.errors import APIError

from src.config import get_settings

logger = logging.getLogger(__name__)

class GeminiService:
    def __init__(self):
        settings = get_settings()
        self.client = genai.Client(api_key=settings.GEMINI_API_KEY)
        self.model_name = "gemini-2.5-flash"
        
    def send_message(self, prompt: str, system_instruction: str, expected_schema: Optional[type] = None) -> Dict[str, Any]:
        """
        Sends a message to Gemini and returns parsed JSON.
        Implements 3x exponential backoff on failure.
        """
        retries = 3
        backoffs = [1, 2, 4]
        
        config_args = {
            "system_instruction": system_instruction,
            "response_mime_type": "application/json"
        }
        
        if expected_schema:
            config_args["response_schema"] = expected_schema

        config = types.GenerateContentConfig(**config_args)

        for attempt in range(retries):
            try:
                response = self.client.models.generate_content(
                    model=self.model_name,
                    contents=prompt,
                    config=config
                )
                
                # Parse JSON
                if response.text:
                    return json.loads(response.text)
                return {}
                
            except APIError as e:
                status_code = getattr(e, "code", 500)
                if status_code == 429 or status_code >= 500:
                    if attempt < retries - 1:
                        logger.warning(f"Gemini API error {status_code}, retrying in {backoffs[attempt]}s...")
                        time.sleep(backoffs[attempt])
                        continue
                logger.error(f"Gemini API error exhausted retries: {e}")
                raise
            except json.JSONDecodeError as e:
                logger.error(f"Failed to parse Gemini JSON output: {e} - Content: {response.text}")
                raise ValueError("Invalid JSON output from model")
                
        raise Exception("Failed to get response from Gemini")

File: backend/src/gemini/gemini.service.ts
/**
 * Gemini AI service — ported from Python src/services/gemini_service.py
 * Uses @google/generative-ai SDK with 3× exponential backoff.
 */
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { GoogleGenerativeAI, ResponseSchema } from '@google/generative-ai';
import { LlmProvider } from '../llm/llm-provider.interface';

@Injectable()
export class GeminiService implements LlmProvider {
  private readonly logger = new Logger(GeminiService.name);
  private readonly genAI: GoogleGenerativeAI;
  private readonly modelName: string;

  constructor(private readonly configService: ConfigService) {
    const apiKey = this.configService.getOrThrow<string>('GEMINI_API_KEY');
    this.modelName =
      this.configService.get<string>('GEMINI_MODEL') ?? 'gemini-2.5-flash';
    this.genAI = new GoogleGenerativeAI(apiKey);
  }

  /**
   * Send a prompt to Gemini and return parsed JSON.
   * Implements 3× exponential backoff on 429/5xx.
   */
  async sendMessage(
    prompt: string,
    systemInstruction: string,
    responseSchema?: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    const retries = 3;
    const backoffs = [1000, 2000, 4000]; // ms

    const model = this.genAI.getGenerativeModel({
      model: this.modelName,
      systemInstruction,
      generationConfig: {
        responseMimeType: 'application/json',
        ...(responseSchema ? { responseSchema: responseSchema as unknown as ResponseSchema } : {}),
      },
    });

    for (let attempt = 0; attempt < retries; attempt++) {
      try {
        const result = await model.generateContent(prompt);
        const text = result.response.text();

        if (text) {
          return JSON.parse(text) as Record<string, unknown>;
        }
        return {};
      } catch (error: unknown) {
        const statusCode = (error as { status?: number }).status ?? 500;

        if (statusCode === 429 || statusCode >= 500) {
          if (attempt < retries - 1) {
            this.logger.warn(
              `Gemini API error ${statusCode}, retrying in ${backoffs[attempt]}ms...`,
            );
            await this.sleep(backoffs[attempt]);
            continue;
          }
        }

        this.logger.error(`Gemini API error exhausted retries: ${error}`);
        throw error;
      }
    }

    throw new Error('Failed to get response from Gemini');
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
